modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,607 @@
+
+        /* Initialize object attributes for the object type */
+        InitializeObjectAttributes(&ObjectAttributes,
+                                   &ObjectPath,
+                                   OBJ_KERNEL_HANDLE,
+                                   NULL,
+                                   NULL);
+
+        /* Open the existing object type by name */
+        Status = ObOpenObjectByName(&ObjectAttributes,
+                                    NULL,
+                                    KernelMode,
+                                    NULL,
+                                    0,
+                                    NULL,
+                                    &ObjectTypeHandle);
+
+        ok_eq_hex(Status, STATUS_SUCCESS);
+        ok(ObjectTypeHandle != NULL, "ObjectTypeHandle is NULL\n");
+
+        if (!skip(Status == STATUS_SUCCESS && ObjectTypeHandle, "No handle\n"))
+        {
+            /* Reference the existing object type by handle */
+            Status = ObReferenceObjectByHandle(ObjectTypeHandle,
+                                               0,
+                                               NULL,
+                                               KernelMode,
+                                               (PVOID)&DummyType,
+                                               NULL);
+
+            ok_eq_hex(Status, STATUS_SUCCESS);
+
+            if (!skip(Status == STATUS_SUCCESS && DummyType, "No pointer to the dummy object's body\n"))
+            {
+                /* Set the procedures for the existing object type */
+                DummyType->TypeInfo.CloseProcedure = CloseProc;
+                DummyType->TypeInfo.OpenProcedure = OpenProc;
+                DummyType->TypeInfo.DeleteProcedure = DeleteProc;
+            }
+
+            /* Close the object type handle */
+            if (ObjectTypeHandle != NULL)
+            {
+                Status = ZwClose(ObjectTypeHandle);
+            }
+        }
+    }
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    return Status;
+}
+
+/*!
+ * @brief This function cleans up resources that were not cleaned up during the tests.
+ */
+static
+VOID
+ObInsert_Cleanup(VOID)
+{
+    NTSTATUS Status;
+
+    /* Make the directory object temporary (which is permanent) if the handle is valid */
+    if (!skip(DirectoryHandle != NULL, "No directory handle\n"))
+    {
+        Status = ZwMakeTemporaryObject(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+
+        /* Close the directory handle */
+        Status = ZwClose(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+    }
+}
+
+/*!
+ * @brief This function tests for successful object insertion.
+ */
+static
+VOID
+ObInsert_Success(VOID)
+{
+    NTSTATUS Status;
+    UNICODE_STRING ObjectName;
+    PDUMMY_TYPE Object;
+    HANDLE Handle;
+    POBJECT_HEADER ObjectHeader;
+
+    /* Initialize the object name */
+    RtlInitUnicodeString(&ObjectName,
```

Comment:
It should be possible to use `UNICODE_STRING ObjectName = RTL_CONSTANT_STRING(L"\\ObInsertTest\\DummyObject_Success");` , here and in the similar cases used elsewhere in this file.